### PR TITLE
TSPS-323 Submit job and make things pretty

### PR DESCRIPTION
**Pull request by mmorgantaylor**
### Description 

Add `submit` command. Behavior while upload in progress:
image

Completed upload/submission:
image

Accompanying PR in main service repo: https://github.com/DataBiosphere/terra-scientific-pipelines-service/pull/157

Other cleanup of CLI command output: 

list pipelines (with and without auth prompt)
image

get info:
image

input validation:
image

### Jira Ticket
https://broadworkbench.atlassian.net/browse/TSPS-323

### CODE REVIEW

File: pyproject.toml
Comment (mmorgantaylor) [inline pyproject.toml:17]: revert this before merging

File: terralab/.terralab-cli-config
Comment (mmorgantaylor) [inline terralab/.terralab-cli-config:2]: revert this before merging

File: CONTRIBUTING.md
Comment (jsotobroad) [inline CONTRIBUTING.md:45]: should this point to `python-client/generated` instead?

also i think these instructions should be in the teaspoons repo in case those get changed (this kind of follows on the question i had on the corresponding PR)

File: CONTRIBUTING.md
Comment (jsotobroad) [inline CONTRIBUTING.md:63]: i think this line can point to the readme at "how to run teaspoons" just in case that changes we dont ahve to update it here too

File: tests/logic/test_submit_logic.py
Comment (jsotobroad) [inline tests/logic/test_submit_logic.py:170]: what are these comments of `Act` and `Arrange` and stuff?

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:65]: should this be in the pipelines_logic file?

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:74]: any particular reason we do typing for some variables and not others?

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:62]: does this work for both relative and absolute pathing? (i assume it does)

File: terralab/commands/submit_commands.py
Comment (jsotobroad) [inline terralab/commands/submit_commands.py:14]: add description for this input

File: terralab/logic/submit_logic.py
Comment (jsotobroad) [inline terralab/logic/submit_logic.py:78]: do we want to present this value to the user in case somethign fails between prepare and start so they know what job id they should try to debug?

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:76]: do we want to call out any extra inputs in the inputs_dict?

**Review by jsotobroad — COMMENTED** [review]
pausing my review sicne we went over a lot of stuff with chris and you probably want to address those first and then i can re-review, let me know what works best with you!

File: terralab/logic/submit_logic.py
Comment (mmorgantaylor) [reply terralab/logic/submit_logic.py:78]: yeah i had this as info and changed it to debug, but can add it back for this reason

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:62]: yes, works for both

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:74]: just cause it's not required in Python and I only sometimes remember to do it

File: tests/logic/test_submit_logic.py
Comment (mmorgantaylor) [reply tests/logic/test_submit_logic.py:170]: they were how some of the autogenerated tests organized the testing code. Arrange is the variables and mocks, Act is actually calling the thing you're testing, and Assert is the assertions. happy to remove them if they're dumb

File: terralab/cli.py
Comment (mmorgantaylor) [reply terralab/cli.py:34]: no, they wanted the command to be `terralab submit [PIPELINE NAME]`, not `terralab pipeline_runs submit [PIPELIEN NAME]`

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:65]: sure 👍

File: terralab/cli.py
Comment (mmorgantaylor) [inline terralab/cli.py:41]: remove this

File: terralab/commands/submit_commands.py
Comment (mmorgantaylor) [inline terralab/commands/submit_commands.py:22]: see if we can instead subclass the commands that call APIs to have this decorator automatically (don't block on this)

File: terralab/commands/pipelines_commands.py
Comment (mmorgantaylor) [inline terralab/commands/pipelines_commands.py:36]: remove initial `\n`

File: terralab/commands/pipelines_commands.py
Comment (mmorgantaylor) [inline terralab/commands/pipelines_commands.py:37]: look into best practices for logging multiline formatted strings

File: terralab/logic/auth_logic.py
Comment (mmorgantaylor) [inline terralab/logic/auth_logic.py:14]: remove initial `\n`

File: terralab/logic/submit_logic.py
Comment (mmorgantaylor) [inline terralab/logic/submit_logic.py:78]: remove initial `\n`

File: tests/logic/test_pipelines_logic.py
Comment (mmorgantaylor) [inline tests/logic/test_pipelines_logic.py:117]: split out warning and failure modes to different tests

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:26]: yea we should make a ticket to add newlines to the end of every `error` log

File: terralab/utils.py
Comment (jsotobroad) [inline terralab/utils.py:41]: Is this `raise` something we're doing with from the caller or should we log an `error` and exit here as well?

File: terralab/commands/pipelines_commands.py
Comment (jsotobroad) [inline terralab/commands/pipelines_commands.py:29]: any particular reason we use tabs here vs a pad?  I can think of some but was just curious.

**Review by jsotobroad — COMMENTED** [review]
some nitpicky comments, looks awesome with the formatting and coloring!!

File: terralab/commands/pipelines_commands.py
Comment (mmorgantaylor) [reply terralab/commands/pipelines_commands.py:29]: the `indented` function actually uses a set number of spaces. the full tab width to my eye looked too big. you could also use `pad_column("", 2)` but `indented` is simpler. don't have strong feelings though if this is confusing.

File: terralab/commands/pipelines_commands.py
Comment (mmorgantaylor) [reply terralab/commands/pipelines_commands.py:40]: i think not. it'll depend on the lengths of the strings in the "first column" which may be different. i suspect it'll largely be determined separately for each place it's used (though i did give the function a default value of 20).

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:26]: yeah that would be nice. will do

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:41]: good point, elsewhere we're logging and error and exiting, and we should be consistent

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:26]: https://broadworkbench.atlassian.net/browse/TSPS-374

File: terralab/commands/pipelines_commands.py
Comment (jsotobroad) [reply terralab/commands/pipelines_commands.py:29]: gotcha that makes sense. we can poke at it as we develop more and get more use cases

File: terralab/utils.py
Comment (jsotobroad) [reply terralab/utils.py:41]: if the message logged is some generic thing it would prob be better to log something more specific here instead

**Review by jsotobroad — APPROVED** [review]
just a comment about the `TypeError` raise.  thanks for all of this, lots of good improvements

File: terralab/utils.py
Comment (mmorgantaylor) [reply terralab/utils.py:41]: it's in line 45: "Input string not parsable to a dictionary."

**Pull request merged**
